PIPELINE: Draft version for the API

ansonwong9695 opened the pull request:
Created 3 endpoints for the front end
- /job: which receive the post request from the front end, use to activate the pipeline
- /job/<jobid>/status : this endpoint allows query about the status on specific job and return the job status to the front end
- /api/v1/pipelines: this endpoint allows front end to know the information of each pipelines

Create 1 endpoints to update the status internally
- /status: which allow the pipeline update the status dictionary when they finished a section of task

Other functionality
-create the clean up function for the status, if job exist more than 20 mins will be deleted.
-pipeline when finished the task it will send the request to the hologram (have not tested it)

ansonwong9695 committed:
PIPELINE: Create the template for the server

ansonwong9695 committed:
PIPELINE: Create the template for the server

UdomkarnBoonyaprasert committed:
PIPELINE: add server functions
-add function for job status
-add function for job request from front end
-update pipeline (nifti2glb) to work with the flask server
-implement job status

UdomkarnBoonyaprasert committed:
PIPELINE: add server functions
-add function for job status
-add function for job request from front end
-update pipeline (nifti2glb) to work with the flask server
-implement job status

ansonwong9695 committed:
PIPELINE: add return status to the front end

ansonwong9695 committed:
PIPELINE: add return status to the front end

UdomkarnBoonyaprasert committed:
PIPELINE: new comp and pl template for NN model
- a prototype of numpy list segmentation component (getMultipleNumpy)
- add new functions to compJobStatus, will also repurpose this comp to a generic http request comp soon
- add template for new pl for abdomen NN model

UdomkarnBoonyaprasert committed:
PIPELINE: add TODO comments for pl
-add initial imports
-add templates
-initial pl draft

ansonwong9695 committed:
add clean up function for server

ansonwong9695 committed:
PIPELINE: Compelete the status cleaning function

-Currently clean up function is done it will delete the jobs in the status if the job
already be done 30 mins ago.

ansonwong9695 committed:
solve the conflict

ansonwong9695 committed:
PIPELINE: add extract zip function on the job endpoint

ansonwong9695 committed:
PIPELINE: create the mapping with post request from the front end

ansonwong9695 committed:
PIPELINE: add Http request component

ansonwong9695 committed:
PIPELINES: passing front end request info to the accessor

nbckr commented:
Good morning :) Thanks for opening the PR, Kawai. Some general remarks:
* When I meant prefix `/api/v1/`, I meant _for all routes_, not just `/pipelines`. You can read more about this stuff [here](https://docs.microsoft.com/en-us/azure/architecture/microservices/design/api-design), but please change to be all routes.
* Please merge `dev` back into your branch and resolve the merge conflicts.
* Please resolve the CodeFactor issues.

nielsboecker commented on HoloPipelines/pipelineController.py:
```diff
@@ -91,5 +95,25 @@ def main():
 
+def startPipeline(jobID, plID, paramList=[],infoForAccessor=[]):
+    configFileName = "pipelineList.json"  # hard coded
+    with open(str(pathlib.Path(newCwd).joinpath(str(configFileName)))) as json_file:
```
semantic naming, maybe just call it `pipelinesConfig`. Actually, why isn't the file called `pipelinesConfig` as well iff it actually contains config?

nielsboecker commented on HoloPipelines/pipelineController.py:
```diff
@@ -91,5 +95,25 @@ def main():
+def startPipeline(jobID, plID, paramList=[],infoForAccessor=[]):
+    configFileName = "pipelineList.json"  # hard coded
+    with open(str(pathlib.Path(newCwd).joinpath(str(configFileName)))) as json_file:
+        lsPipe = json.load(json_file)
```
What does "ls" stand for? why abbreviate?

nielsboecker commented on HoloPipelines/pipelineController.py:
```diff
@@ -91,5 +95,25 @@ def main():
     json_file.close()
 
 
+def startPipeline(jobID, plID, paramList=[],infoForAccessor=[]):
```
Please make sure to run `black` / the automatic git hooks to ensure a consistent formatting.

nielsboecker commented on HoloPipelines/pipelineController.py:
```diff
@@ -91,5 +95,25 @@ def main():
+    configFileName = "pipelineList.json"  # hard coded
+    with open(str(pathlib.Path(newCwd).joinpath(str(configFileName)))) as json_file:
+        lsPipe = json.load(json_file)
+        print(str(["python", lsPipe[plID]["src"], jobID] + paramList+[infoForAccessor]))
```
Please no "print" in productive code; if it is still needed, use `logging`

nielsboecker commented on HoloPipelines/pipelineController.py:
```diff
@@ -91,5 +95,25 @@ def main():
+    with open(str(pathlib.Path(newCwd).joinpath(str(configFileName)))) as json_file:
+        lsPipe = json.load(json_file)
+        print(str(["python", lsPipe[plID]["src"], jobID] + paramList+[infoForAccessor]))
+        #subprocess.run(["python", lsPipe[plID]["src"], jobID] + paramList+infoForAccessor, cwd=newCwd)
```
Why is this commented out? Isn't this starting the actual processing?

nielsboecker commented on HoloPipelines/pipelineController.py:
```diff
@@ -91,5 +95,25 @@ def main():
+        lsPipe = json.load(json_file)
+        print(str(["python", lsPipe[plID]["src"], jobID] + paramList+[infoForAccessor]))
+        #subprocess.run(["python", lsPipe[plID]["src"], jobID] + paramList+infoForAccessor, cwd=newCwd)
+        # currently pipelineNifti2glb pipelineDicom2glb are the only pipelines that can take jobID
```
Hmm which other pipelines need to be changed? Can you please create tasks on the taskboard so we don't lose track?

nielsboecker commented on HoloPipelines/pipelineController.py:
```diff
@@ -91,5 +95,25 @@ def main():
+    json_file.close()
+
+
+def getPipelineList():
```
If you have this method, why don't you use it in line 100? Also, again, "ls" and "Pipe" are both unnecessary abbreviations in the variable name

boonwj committed:
PIPELINE: Remove dicom zip file

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+from zipfile import ZipFile
+
+app = Flask(__name__)
+app.config["JSON_ADD_STATUS"] = False
```
What is `JSON_ADD_STATUS` doing? The name is not expressive. Also to me it seems that this variable is never accessed?

ansonwong9695 replied:
because I use the flask_json library for creating json response to the front end and in default it will generate a status code by default but the status code i want to generate is not the default value. for example in default  library will return a json response to the request with status code 200 but i want to send 202 instead.

nielsboecker replied:
Ohh okay. Please add flask_json to the requirements.txt then, along with other dependencies. 

Btw, if i understand the docs correctly, it will just include the status code _that you set_, not always the default value. If you do `return json_response(jobID=jobID, status_code=202)`, it should include 200. But it's fine to just turn the feature off, I guess.

ansonwong9695 committed:
PIPELINE: merge with dev

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+
+# global variables
+this_cwd = pathlib.Path.cwd()
+save_directory = Path("input")
```
Could be called `input_directory` for consistency

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+def save_request_file_to_input_dir(filename, request_input_data_URL):
+    global save_directory
+    if not save_directory.is_dir():
+        os.mkdir("input")
```
Shouldn't hard-code `"input"` here, use the method argument

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+        os.mkdir("input")
+    response = requests.get(request_input_data_URL)
+    if response.status_code != 200:
+        return json_response(jobID="image study endpoint failure", status_code=404)
```
Status code 404 is okay, but the response should have a different format. There is no jobId.
```
{
    "message": "Failed to fetch imaging study from PACS"
}
```

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+logging.basicConfig(level=logging.INFO, format=FORMAT)
+
+
+def save_request_file_to_input_dir(filename, request_input_data_URL):
```
Could be renamed to `fetch_request_input_file` to make it more expressive

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+    file_dir = this_cwd.joinpath(str(save_directory), filename)
+    path_to_store_unzip_file = this_cwd.joinpath(str(save_directory), filename_unzip)
+    with ZipFile(str(file_dir), "r") as zipObj:
+        # unzip
```
The comments are not necessary here

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+
+
+# cleaning the status
+@app.before_first_request
```
Instead of using this decorator, could you just call this from your `main` method?

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+    def run_job():
+        while True:
+            global status
+            # logging.info("status before loop: " +str(status))
```
Please remove all the commented log statements. However, it would be good to have one debug statement within the task that runs every 30 sec, that contains some useful information on the current state.

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+            global status
+            # logging.info("status before loop: " +str(status))
+
+            for job in status.copy():
```
why do you `copy()`? This seems like an unneccessary workload and runs every 30 sec...

ansonwong9695 replied:
because i am modifying the status dict inside a loop so i need to create a copy of the status

nielsboecker replied:
Sorry I don't understand. Why do you have to copy it then? What's the problem with the loop? What will break if you do ` for job in status: ...`

ansonwong9695 replied:
RuntimeError: dictionary changed size during iteration
this is the error i will get if i do not use copy, because in this for loop i will delete the key and value inside this dictionary so the size is going to change

nielsboecker replied:
Hm not sure, will leave it for now as the dict will be replaced most likel

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+
+# update the status from pipeline
+@app.route("/status", methods=["POST"])
+def getTheStatus():
```
Please make sure to use snake_case consistently, not camelCase

nielsboecker replied:
Why is the method called "get"? If this is used to update, give it an appropriate name like `update_job_status`

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+    global status
+    current_job_status = request.get_json()
+    status.update(current_job_status)
+    return json.dumps(status)
```
This method would benefit from an informational debug statement

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+    global status
+    current_job_status = request.get_json()
+    status.update(current_job_status)
+    return json.dumps(status)
```
Does it need to always return the whole status for all jobs?? Maybe better just the current one

ansonwong9695 replied:
yes, it only returns the current one

nielsboecker replied:
Oh so you mean you changed it? Because the code as it currently is commited here looks like you return `status` (all) instead of `current_job_status`

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+def send_list_of_pipelines():
+    global pipeline_list
+    pipeline_dict = {}
+    for (
```
Why the line breaks and the parenthesis?
Usual syntax is `for key, value in dict.items():`
Also, in your case, instead of `key` and `value`, use `plid` and `pipeline` or something like that.

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+    ) in (
+        pipeline_list.items()
+    ):  # not complete. the value for inputConstraints is wrong (atm it's just number of param)
+        pipeline_dict[key] = {
```
A couple of things here:
1. You are doing a mapping here. Could extract that to a method like `map_pipelines_info()`.
1. You are doing this in every request to this endpoint, although the pipeline list is static. You could just call it once on the top level.
1. Actually the best thing might be to just _change the format of the pipeline list_ so that you don't have to map it. Just have the 1:1 same keys in the pipeline list json. Then your mapping becomes real easy. It is basically a filtering for the values you want. You send them unchanged, and you jjust use your mapping method to exclude internal fields. See https://stackoverflow.com/a/3420156/8495954

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+            "outputExampleImageUrl": value["outputExampleImageUrl"],
+        }
+
+    return json.dumps(pipeline_dict)  # should we str() here?
```
No we don't need `str()`, because that's exactly what `json.dumps()` provides (see https://docs.python.org/3/library/json.html) Please remove comment.

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+
+# use to start the pipeline
+@app.route("/job", methods=["POST"])
+def send_job_start_response():
```
The name should reflect the core thing this method does, for instance `start_job`

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+
+def unzip_the_file(filename):
+    # get unzip input filename
+    filename_unzip = filename.rsplit(".", 1)[0]
```
Do you need the second parameter here?

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+        filename = request_input_data_URL.rsplit("/", 1)[1]
+        logging.info("filename: " + filename)
+
+    save_request_file_to_input_dir(filename, request_input_data_URL)
```
This method can just return the `response.content` rather than writing it to a file. Then use that as a paramter to the unzip method, and you have done it in-memory and avoided an additional write to disk: https://stackoverflow.com/questions/2463770/python-in-memory-zip-library

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+    # create output dir
+    global output_directory
+    if not output_directory.is_dir():
+        os.mkdir("output")
```
Don't hardocde the string, why do you have the variable?

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+    jobID = str(uuid.uuid1())
+    global jobid2plid
+    jobid2plid.update({jobID: request_plid})
+    if request_plid != "lungDicom2glb":
```
Duplicated code. Use the same variable for both. Then, do a positive matching, that means for any of the pipelines that _DO_ have the additional param, add it.

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+    logging.info("arglist: " + str(arglist))
+
+    #create a list that fetch the response info that needs to post with hologram to the accessor (checking is missing)
+    info_for_accesor={
```
choose a semantic name that is decoupled from other services, e.g. `meta_data`. To assemble the data, you can do the trick I mentioned earlier: https://stackoverflow.com/a/3420156/8495954

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+    if jobid in status:
+        status_for_current_jobid = {jobid: status[jobid]["status"]}
+    else:
+        status_for_current_jobid = {jobid: "does not exist"}
```
Should not use `jobId`, but `message` as key. Also, do you set the reponse status? should be 200 or 404 respectively.

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,201 @@
+
+
+if __name__ == "__main__":
+    app.run(debug=True, host="127.0.0.1")
```
Shoud not be debug

nielsboecker commented on HoloPipelines/pipelines/NOTDONEpipelineAbdomenDicom2glb.py:
```diff
@@ -0,0 +1,32 @@
+# ----this pl in a nutshell:
```
Why is this file here? It seems liike it is only a draft, not ready to be merged into `dev`. Also doesn't it belong into #32? Maybe you should revert it for now, as I don't want to have it in the `dev` branch like this, and it doesn't seem related with this PR much.

nielsboecker replied:
deleted

nielsboecker commented on HoloPipelines/pipelines/components/compConvertDicom2nifti.py:
```diff
@@ -0,0 +1,11 @@
+import numpy as np
```
Why is this file in the PR? Does it have anything to do with the server?

nielsboecker commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,73 @@
+from pathlib import Path
+from os import path
+
+def sendFilePostRequest(url, inputFile, outputFile):
```
Please use snake_case for all methods and parameter names.

nielsboecker replied:
Also please run `black` on all files to format (so there is a space between arguments etc.). I highly recommend using an IDE like PyCharm, it will show you such issues while you code.

nielsboecker commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,73 @@
+def sendFilePostRequest(url, inputFile, outputFile):
+    inputFile = str(pathlib.Path(inputFile))
+    outputFile = str(pathlib.Path(outputFile))
+    file = {"file": open(inputFile, "rb")}
```
use `with open...` for the file handling

nielsboecker replied:
Also, use more specific names like `input_file`. Don't use the same variable for something else in line 12.

nielsboecker commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,73 @@
+    file.write(response.content)
+    file.close()
+    returnCode = response.status_code
+    print(returnCode)
```
Use logging. Also you may not need a `returnCode` variable, just use the `response.status_code` value inline

nielsboecker commented on HoloPipelines/pipelines/components/compJobStatus.py:
```diff
@@ -0,0 +1,39 @@
+    print(returnCode)
+
+
+def sendFilePostRequest(url, inputFile, outputFile):
```
Duplicate code from `compHttpRequest`?

nielsboecker replied:
Also, the method is never called? The other one in `compHttpRequest` is also never called?

nielsboecker commented on HoloPipelines/pipelines/components/getMultipleNumpy.py:
```diff
@@ -0,0 +1,50 @@
+#TODO change file name cos i legit dont know what to name this. File naming is not the same to the rest of comp format anyway
```
Renamed to `HoloPipelines/pipelines/components/compSeparateNumpy.py ` in Pap's branch - you guys should probably merge your stuff or you will get into trouble if you each make changes in now different files

nielsboecker replied:
deleted

nielsboecker reviewed (changes requested):
As discussed

ansonwong9695 committed:
PIPELINE: change the accessor post method mapping

nielsboecker commented on HoloPipelines/pipelines/components/test_accessor.py:
```diff
@@ -0,0 +1,25 @@
+from compHttpRequest import sendFilePostRequestToAccessor
```
As this is a test, it should probably be in the `test` folder

nielsboecker commented on HoloPipelines/pipelines/components/test_accessor.py:
```diff
@@ -0,0 +1,25 @@
+from compHttpRequest import sendFilePostRequestToAccessor
+
+url="http://localhost:3200"
+directory="/home/kawai/git/HoloRepository-Core/HoloPipelines/output/left-scfe-pelvis-bone.glb"
```
Hard-coded, this won't work on any other machine. Same for the localhost -- when this test runs on the build server, it will fail because there is nothing at `localhost:3200`. Also you will repeatedly create actual new data in the database if you don't mock, but do a real call to the Accessor.

Not sure what to do with this. Any ideas @boonwj? Maybe [skip the test in automatic runs](https://stackoverflow.com/questions/38442897/how-do-i-disable-a-test-using-py-test), rename it to `accessor_integration_test`, and add documentation that this can only be run manually.

nielsboecker commented on HoloPipelines/pipelines/components/test_accessor.py:
```diff
@@ -0,0 +1,25 @@
+  }
+
+
+sendFilePostRequestToAccessor("pelvis generate from dicom to glb",url,directory,"pelvis bone gernerate from dicom to glb pipeline","hip","2017-07-15T15:20:25Z","2017-07-15T15:20:25Z","pelvis generate from dicom2glb",author,patient)
```
This test is not really testing anything. No asserts, just a shot-in-the-dark method call. In case we keep this test around as a manual integration test or whatever: You could return the response code from the mmethod, and in your test verify that it was successful.

(or better, return the _body_ of the response. You know, a successful POST to the Accessor returns the JSON of the created hologram. then you can actually verify that it created the right data, same aid, name, etc)

nielsboecker commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,73 @@
+from pathlib import Path
+from os import path
+
+def sendFilePostRequest(url, inputFile, outputFile):
```
Is this dead code? Seems you never call it.

nielsboecker commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,73 @@
+
+def sendFilePostRequestToAccessor(title,url, outputFileDir,description,bodySite,dateOfImaging,creationDate,creationDescription,author,patient):
+    outputFileDir = str(pathlib.Path(outputFileDir))
+    sizeOfOutputFile=str(path.getsize(outputFileDir)/1000)
```
Should be 1024?

nielsboecker commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,73 @@
+
+
+def sendFilePostRequestToAccessor(title,url, outputFileDir,description,bodySite,dateOfImaging,creationDate,creationDescription,author,patient):
+    outputFileDir = str(pathlib.Path(outputFileDir))
```
Again, please use snake_case for variable names. You should install an IDE or some plugin for your text editor so that it does a check against PEP8 style guide and warns you before you commit.

nielsboecker commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,73 @@
+    outputFileDir = str(pathlib.Path(outputFileDir))
+    sizeOfOutputFile=str(path.getsize(outputFileDir)/1000)
+    file = {"file": open(outputFileDir, "rb")}
+    #print("file: "+str(file)
```
Pls remove

nielsboecker commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,73 @@
+    }
+
+
+    requestBody = {
```
This will need to be changed to form data and multifor , but I think you are already working on that.

nielsboecker commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,73 @@
+    response = requests.post(url, data=requestBody)
+
+    returnCode = response.status_code
+    print(returnCode)
```
Use logging instead, and maybe provide some context instead of just the number. Also, the response's body is interesting as well.

nielsboecker commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,73 @@
+
+    returnCode = response.status_code
+    print(returnCode)
+    return requestBody
```
Why would you return the requestBody? I think you either want to return a boolean (whether the response was successful) ~~or maybe the response body (but that might be unnecessary if you don't acually use it anywhere)~~.

nielsboecker commented on HoloPipelines/pipelines/components/compJobStatus.py:
```diff
@@ -0,0 +1,39 @@
+import requests
```
I don't like that every component that calls this component has their own, free-style strings to set their status. I would prefer if you had either an enum or a couple of constant string variables here in this file, that other components reference for the `jobStatus` argument)

nielsboecker commented on HoloPipelines/pipelines/components/compJobStatus.py:
```diff
@@ -0,0 +1,39 @@
+
+def updateStatus(
+    jobID, jobStatus
+):  # TODO rename this comp to just http request? and rename this function to updateJobStatus
```
No, it is okay to have status in one component. You could have another component that just does the http requests. And in the status component, you just call the other one (so you don't even know the pipeline server url here in this one)

nielsboecker replied:
But if this component here is responsible for keeping the status, then you should also move the dict of the status inside this file. And then from other files, you can import it to access it.

nielsboecker commented on HoloPipelines/pipelines/components/compJobStatus.py:
```diff
@@ -0,0 +1,39 @@
+def updateStatus(
+    jobID, jobStatus
+):  # TODO rename this comp to just http request? and rename this function to updateJobStatus
+    pipelineServerURL = "http://localhost:5000/status"
```
Please change 5000 to 3100 everywhere, as discussed. Ideally, it should be configurable and read from a separate config file.

nielsboecker replied:
Please add documentation why we are doing this. And maybe rename to `self_url` to make it more clear what is going on.

nielsboecker commented on HoloPipelines/pipelines/components/compJobStatus.py:
```diff
@@ -0,0 +1,39 @@
+):  # TODO rename this comp to just http request? and rename this function to updateJobStatus
+    pipelineServerURL = "http://localhost:5000/status"
+    data = {jobID: {"status": jobStatus, "timestamp": str(datetime.now())}}
+    # data = {"jobID": "j0", "status": "hey", "stamp": "2020"}
```
Pls remove

nielsboecker commented on HoloPipelines/pipelines/components/compJobStatus.py:
```diff
@@ -0,0 +1,39 @@
+    # data = {"jobID": "j0", "status": "hey", "stamp": "2020"}
+    response = requests.post(url=pipelineServerURL, json=data)
+    returnCode = response.status_code
+    print(returnCode)
```
Again, please use logging, and write some context to it. Also no need to store it in a separate variable.

nielsboecker commented on HoloPipelines/pipelines/components/compJobStatus.py:
```diff
@@ -0,0 +1,39 @@
+    print(returnCode)
+
+
+def sendGetRequest(url, requestBody):
```
1. I think is never used as well??
1. Please put only status related methods inside this file and remove unrelated stuff.

nielsboecker commented on HoloPipelines/pipelines/components/compJobStatus.py:
```diff
@@ -0,0 +1,39 @@
+def sendGetRequest(url, requestBody):
+    response = requests.post(url=url, json=requestBody)
+    returnCode = response.status_code
+    print(returnCode)
```
Print statement

nielsboecker commented on HoloPipelines/pipelines/components/compJobStatus.py:
```diff
@@ -0,0 +1,39 @@
+
+
+if __name__ == "__main__":
+    updateStatus("j0", "hello")  # TODO remove once done testing
```
😐

nielsboecker replied:
This should not be here. Instead you can write an actual test for this component (after moving the actual state here), setting state , and then verifying that it is set. Easy as that.

nielsboecker commented on HoloPipelines/pipelines/components/compJobStatus.py:
```diff
@@ -0,0 +1,39 @@
+    updateStatus("j0", "hello")  # TODO remove once done testing
+    print(
+        "You shouldn't be able to run this component directly after we're done testing"
+    )  # TODO look at this too pls
```
What does this mean? Such comments should be removed before merging this to `dev`

nielsboecker replied:
Was marked as resolved but comment still here. I will remove it now.

nielsboecker commented on HoloPipelines/pipelines/components/compHttpRequest.py:
```diff
@@ -0,0 +1,73 @@
+    return outputFile
+
+
+def sendFilePostRequestToAccessor(title,url, outputFileDir,description,bodySite,dateOfImaging,creationDate,creationDescription,author,patient):
```
These are too many arguments. Consider redisgning it or [introducing a parameter object](https://softwareengineering.stackexchange.com/a/145066)

nielsboecker commented on HoloPipelines/pipelines/pipelineLungDicom2glb.py:
```diff
@@ -1,14 +1,18 @@
 
 
-def main(dicomPath, outputGlbPath):
+def main(jobID, dicomPath, outputGlbPath,infoForAccessor):
```
`metaData` instead of `infoForAccessor`

nielsboecker commented on HoloPipelines/pipelines/pipelineLungDicom2glb.py:
```diff
@@ -19,19 +23,35 @@ def main(dicomPath, outputGlbPath):
         deleteOriginalObj=True,
         compressGlb=False,
     )
     print("lungDicom2glb: done, glb saved to {}".format(generatedGlbPath))
```
Use logging instead of print

nielsboecker commented on HoloPipelines/pipelines/pipelineLungDicom2glb.py:
```diff
@@ -19,19 +23,35 @@ def main(dicomPath, outputGlbPath):
     print("lungDicom2glb: done, glb saved to {}".format(generatedGlbPath))
+    compJobStatus.updateStatus(jobID, "Finished")
+    compHttpRequest.sendFilePostRequestToAccessor(
+        infoForAccessor["bodySite"]+"apply on lung segmentation",
```
Missing a space here. Also the body site is pretty obvious I guess :) How about just "Lung segmentation"?

nielsboecker commented on HoloPipelines/pipelines/pipelineLungDicom2glb.py:
```diff
@@ -19,19 +23,35 @@ def main(dicomPath, outputGlbPath):
+    compJobStatus.updateStatus(jobID, "Finished")
+    compHttpRequest.sendFilePostRequestToAccessor(
+        infoForAccessor["bodySite"]+"apply on lung segmentation",
+        "http://localhost:3200/api/v1/holograms",
```
The URL should not be hard-coded; it could be another domain, once we deploy to production. So please use a config file. Also, the URL should not be aparameter for this method, because it's always going to the accessor. It should be kept inside the http component

nielsboecker commented on HoloPipelines/pipelines/pipelineLungDicom2glb.py:
```diff
@@ -19,19 +23,35 @@ def main(dicomPath, outputGlbPath):
+        infoForAccessor["description"],
+        outputGlbPath,infoForAccessor["bodySite"],
+        infoForAccessor["dateOfImaging"],
+        datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
```
creationDate should not be a paramter for this method, as it will be dynamically created all the same. Should be done in http component.

nielsboecker commented on HoloPipelines/pipelines/pipelineLungDicom2glb.py:
```diff
@@ -19,19 +23,35 @@ def main(dicomPath, outputGlbPath):
+        outputGlbPath,infoForAccessor["bodySite"],
+        infoForAccessor["dateOfImaging"],
+        datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
+        "Generate glb mesh from dicom",
```
For consistency, how about having the parameter just be the pipeline name? Then in the http component, the creationDescription could always be `f"Generated with {pipelineName] pipeline."`

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,202 @@
+
+
+# update the status from pipeline
+@app.route("/api/v1/status", methods=["POST"])
```
There are actually more elegang ways to have the same prefix for all routes, which also makes it easier if it changes. See https://stackoverflow.com/questions/18967441/add-a-prefix-to-all-flask-routes

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,202 @@
+from flask import Flask, request
```
This file does way too much stuff which is not server related. Please do at least extract other components:
* to keep the status of jobs
* to do the downloading and unzipping stuff

nielsboecker replied:
Rather than having them on the top level, could you move this file and `pipelineController` to a new directory/module `HoloPipelines/core`

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,202 @@
+output_directory = Path("output")
+pipeline_list = getPipelineList()
+status = {
+    "j0": {"status": "segment", "stamp": "2019-08-05 14:09:19"},
```
Ummm??? This should probably be removed?? You want to start with an empty status dictionary.

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,202 @@
+    "j0": {"status": "segment", "stamp": "2019-08-05 14:09:19"},
+    "j1": {"status": "segment", "stamp": datetime.now().strftime("%Y-%m-%d %H:%M:%S")},
+}
+jobid2plid = {"j0": "nifti2glb", "j1": "dicom2glb"}
```
None of these three dictionaries are needed. job to plid, job to inputURL, job to outputURL. Please remove all.

nielsboecker reviewed (changes requested):
Please follow up with the comments, feel free to ask if you have any questions. Before we can accept the PR, we should also verify that
- the status updates are actually working (can be verified if you add debug statements)
- the data that gets created for a job will also be cleaned after it's finished
- the POST to the Accessor is actually working

nbckr commented:
@ansonwong9695, @UdomkarnBoonyaprasert: I have some other TODOs for you guys; we can track them in this post and cross them out when you finished them:
1. I see an architectural problem: Initially, we said both the data fetching from PACS and the data posting to Accessor should be their own components, that can be part of each pipeline. In your code however, the fetching gets done right in the server, whereas the posting gets done at another level (in each pipeline). Please refactor.
1. You need to clean up after yourselves. What happens with all the input files you download after the job is finished? They will stay there forever, currently. Also they will all be in the same directory, it is impossible to keep track what came from where. Potentially there can even be issues if different pipelines try to download a file with the same name at the same time. How about creating a folder `jobs/<jobid>` for each job. `jobs/<jobid>/input` for your stuff from the PACS and `jobs/<jobid>/output` for what you are generating; and then maybe deleting that once stuff gets removed from the status? @boonwj any other ideas?
1. You should add flask to your `dependencies.txt`
1. You should update the `README` on how to start the server and document the endpoints.
1. There are two empty files, please remove: pipelines/components/compStoreGLB and pipelines/components/compOutputGLB
1. You should add a Dockerfile (this can be a separate PR, but don't forget it)
1. There should be more error handling (you always expect that all fields are coming exactly as expected, but what happens otherwise?) - can be a separate PR

ansonwong9695 committed:
PIPELINE: fix the bug fot posting to accesor

ansonwong9695 committed:
PIPELINE: remove the temporary testing script for accesor

ansonwong9695 committed:
PIPELINE: change the class to dictionary for accessor info

ansonwong9695 committed:
PIPELINE: change it to camel cases

UdomkarnBoonyaprasert committed:
PIPELINE: add new component for downloading images

ansonwong9695 committed:
PIPELINE: make change according to commendts

ansonwong9695 committed:
PIPELINE: resolve conflict

UdomkarnBoonyaprasert committed:
PIPELINE: add cleanup component

UdomkarnBoonyaprasert committed:
PIPELINE: solve conflict

UdomkarnBoonyaprasert committed:
PIPELINE: fix bug in pipelineServer

UdomkarnBoonyaprasert committed:
PIPELINE: add cleanup for lung pipeline

UdomkarnBoonyaprasert committed:
PIPELINE: add temp file cleanup for nifti2glb pipeline
- add cleanup for pipelineNifti2glb (only of the temp obj file)
- delete pipelineNifti2obj

UdomkarnBoonyaprasert committed:
PIPELINE: update requirements.txt
add Flask and Flask-JSON to dependencies

ansonwong9695 committed:
PIPELINE: more change according to comments

ansonwong9695 committed:
PIPELINE: fix conflict

nielsboecker committed:
Merge branch 'PIPELINE/API_file_cleanup' into PIPELINE/API

This was a tough one. I have no idea if this broke something.

nielsboecker committed:
PIPELINE: Delete unused stuff

nielsboecker commented on HoloPipelines/pipelines/components/compStatus.py:
```diff
@@ -0,0 +1,9 @@
+from datetime import datetime
+
+status = {
```
Why are there still things in here? Shouldn't this start out empty???

nielsboecker replied:
I'll fix it in refactoring branch.

nielsboecker commented on HoloPipelines/pipelines/components/compJobPath.py:
```diff
@@ -0,0 +1,35 @@
+    return str(pathlib_job_path)
+
+
+def make_str_job_path(job_ID, sub_dir_list, create_sub_directories=True):
```
Method name implies that it just returns a string, but it creates directories

nielsboecker commented on HoloPipelines/pipelines/components/compJobPath.py:
```diff
@@ -0,0 +1,35 @@
+this_comp_path = str(pathlib.Path(str(os.path.dirname(os.path.realpath(__file__)))))
+
+
+def pathlib_job_path(job_ID, create_dir=True):
```
Why "pathlib" in method name?

nielsboecker commented on HoloPipelines/pipelines/components/compJobPath.py:
```diff
@@ -0,0 +1,35 @@
+    return job_path
+
+
+def str_job_path(job_ID):
```
Unused parameter `job_ID`

nielsboecker commented on HoloPipelines/pipelines/components/compJobPath.py:
```diff
@@ -0,0 +1,35 @@
+    return str(pathlib_job_path.joinpath(*sub_dir_list))
+
+
+def clean_up(job_ID):
```
Unused parameter `job_ID` -- This can never work. Have you seen this work for any actual job???

nielsboecker commented on HoloPipelines/pipelines/components/compJobPath.py:
```diff
@@ -0,0 +1,35 @@
+    return str(pathlib_job_path)
+
+
+def make_str_job_path(job_ID, sub_dir_list, create_sub_directories=True):
```
Unused parameter `job_ID` --- How can this even work???

nielsboecker replied:
Again, let me advise you strongly to use an IDE, as it helps:
![image](https://user-images.githubusercontent.com/11090412/63204720-b40cb280-c092-11e9-9e13-fac562af3444.png)

nielsboecker commented on HoloPipelines/pipelines/components/compJobPath.py:
```diff
@@ -0,0 +1,35 @@
+
+
+def clean_up(job_ID):
+    if os.path.exists(str_job_path):
```
`str_job_path` is the method name, but you are not calling it?

nielsboecker commented on HoloPipelines/pipelines/components/compJobPath.py:
```diff
@@ -0,0 +1,35 @@
+
+
+def str_job_path(job_ID):
+    return str(pathlib_job_path)
```
`pathlib_job_path` is the method name, but you are not calling it?

nielsboecker replied:
Why would you even need a method that just does a str() of something? Also why don't you just return a str() of thhe Path in line 13?

nielsboecker commented on HoloPipelines/pipelines/components/compJobPath.py:
```diff
@@ -0,0 +1,35 @@
+
+def make_str_job_path(job_ID, sub_dir_list, create_sub_directories=True):
+    if (
+        not os.path.isdir(str(pathlib_job_path.joinpath(*sub_dir_list).parent))
```
In line 22-26 you again use `pathlib_job_path` like a variable, but it is a method,, and you are not calling it? Also I don't understand what is happening iin these lines. Please clarify and/or write comments.

nielsboecker commented on HoloPipelines/pipelines/components/compGetInput.py:
```diff
@@ -0,0 +1,31 @@
+from zipfile import ZipFile
+import logging
+
+input_directory = Path("input")
```
How is this supposed to be the job-specific dir??

nielsboecker replied:
I'll fix it in refactoring branch.

nielsboecker commented on HoloPipelines/pipelineConfig.py:
```diff
@@ -0,0 +1,140 @@
+        logging.debug("This is logging for TEST1")
+
+
+def dynamicLoadingPipeline(plID, parameter_dict):
```
This method is duplicate here  and in pipelineController and both seem to be unused? @ansonwong9695

nielsboecker replied:
I'll fix it in refactoring branch.

nielsboecker commented on HoloPipelines/pipelineController.py:
```diff
@@ -40,56 +48,93 @@
     json_file.close()
 
 
+def startPipeline(pipeline_ID, parameter_dict):
```
Duplicate code and seems to never be called??

nielsboecker replied:
I'll fix it in refactoring branch.

nielsboecker commented on HoloPipelines/pipelineController.py:
```diff
@@ -40,56 +48,93 @@
+    return list_of_pipeline
+
+
+def dynamicLoadingPipeline(plID, parameter_dict):
```
Duplicate code and seems to never be called?

nielsboecker replied:
I'll fix it in refactoring branch.

nielsboecker commented on HoloPipelines/pipelineServer.py:
```diff
@@ -0,0 +1,108 @@
+# global variables
+this_cwd = pathlib.Path.cwd()
+
+output_directory = Path("output")
```
How is this supposed to be the job specific directory?

nielsboecker replied:
I'll fix it in refactoring branch.

nielsboecker committed:
PIPELINES: Clean up various code smells

nielsboecker committed:
PIPELINE: Fix CodeFactor issues

nbckr commented:
[CodeFactor](https://www.codefactor.io/repository/github/nbckr/holorepository-core) found an issue: Duplicate Code

It's currently on:
[HoloPipelines\pipelineConfig.py:29-82
](https://github.com/nbckr/HoloRepository-Core/blob/d23976d68e06445195e559343b1c6743b133cec3/HoloPipelines/pipelineConfig.py#L29-L82)[HoloPipelines\pipelineController.py:26-79
](https://github.com/nbckr/HoloRepository-Core/blob/d23976d68e06445195e559343b1c6743b133cec3/HoloPipelines/pipelineController.py#L26-L79)Commit d23976d68e06445195e559343b1c6743b133cec3

nielsboecker commented on HoloPipelines/pipelineConfig.py:
```diff
@@ -0,0 +1,140 @@
+    default="pipelineList.json",
+    type=str,
+    help="path to pipeline config file relative to pipelineConfig",
+)
```
Duplicate Code

It's currently on:
HoloPipelines\pipelineConfig.py:29-82
HoloPipelines\pipelineController.py:26-79

nielsboecker replied:
I'll fix it in refactoring branch.

nielsboecker reviewed (changes requested):
@ansonwong9695 Thanks for cleaning up.
* Most things are okay now; i cleaned up some small stuff myself
* Please go through the few comments relevant for your code sections
* Please confirm that everything stll works (or fix if i introduced small bugs)
* ~~We will push the Dockerfile and the logging to files feature to later (not in this PR), i created stories on the task board~~
* ~~there are some issues in CodeFactor left; I think we can leave them there as these components wil hopefully be refactored~~

@UdomkarnBoonyaprasert You told me the clean-up functionality was finished, so now I am thoroughly confused. The clean up component can never have worked, because of crucial errors in there?
* please fix these and all other comments
* please verify and confirm here that the clean up actually worked for you
* please verify that my merge hasn't broken your code
* ~~ideally the clean up should have tests but there is no time, so screw that~~

@~~UdomkarnBoonyaprasert and @ansonwong9695: There are also conflicts for many files from `dev`, so that's another thing we will have to fix...~~ Done by me.

nielsboecker committed:
Merge branch 'dev' into PIPELINE/API

nielsboecker committed:
PIPELINE: Fix typos

nielsboecker committed:
PIPELINE: Fix requirements.txt

nielsboecker committed:
PIPELINE: Fix imports and other issues

nbckr commented:
Sorry @ansonwong9695, I realised that the build for your work was broken all along. I fixed several issues with the imports, but the tests still fail. I think  it is because you changed the `main` method signature (using the object to bundle all the arguments). But you did not change the tests to reflect that. Can you please update the tests?

nielsboecker committed:
PIPELINE: Disable broken tests

nielsboecker committed:
PIPELINE: Fix broken arg name

nbckr commented:
**Update**: I found that the tests fail for different reasons actually. It's a big, big, mess to be honest. I disabled the tests for now, and you can fix them later (in a different branch!)

nbckr commented:
Note that there are still many issues in this branch at the time of merging, especially in "compJobPath". I am merging it now to prevent future merge conflicts, as I am doing a heavy refactoring in #45 . In that branch I also fix most issues. The "compJobPath" issues can be solved in #59.
